src/amqp-client/connection.cr

```diff
@@ -225,16 +226,20 @@ class AMQP::Client
       io.read_timeout = nil
     end
 
-    private def self.start(io, user, password, name)
+    private def self.start(io, user, password, connection_information, name)
       io.write AMQ::Protocol::PROTOCOL_START_0_9_1.to_slice
       io.flush
       Frame.from_io(io) { |f| f.as?(Frame::Connection::Start) || raise Error::UnexpectedFrame.new(f) }
+      product_version = connection_information.try(&.product) ? (connection_information.try(&.product_version) || nil) : AMQP::Client::VERSION
+      platform_version = connection_information.try(&.platform) ? (connection_information.try(&.platform_version) || nil) : Crystal::VERSION
+
       props = Arguments.new({
-        connection_name: name,
-        product:         "amqp-client.cr",
-        platform:        "Crystal",
-        version:         AMQP::Client::VERSION,
-        capabilities:    Arguments.new({
+        connection_name:  name,
+        product:          connection_information.try(&.product) || "amqp-client.cr",
+        platform:         connection_information.try(&.platform) || "Crystal",
+        product_version:  product_version,
+        platform_version: platform_version,
+        capabilities:     Arguments.new({
```

Comment:
Instead of having default values here and all `try` and `||` calls, what  about setting the default values in the `ConnectionInformation` record definitions? Then you can make the properties non-nilable and you don't have to do any nil checks and such here.

```crystal
record(ConnectionInformation,
  product = "amqp-client.cr",
  product_version = AMQP::Client::VERSION,
  platform = "Crystal",
  platform_version = Crystal::VERSION) do
    property product, product_version, platform, platform_version
  end
```

Reply:
We could do that, but not allowing the properties to be nilable forces LavinMQ to know that it must send `product_name` and `product_version`, otherwise one might send `product_name` without `product_version`, and `product_version` will have default value and be wrong in combination with product name.